### ISSUE TITLE
can't get my app to compile when using v0.17.1 of the package

### ISSUE THREAD

**Issue by darre1** [luanpotter/audioplayers#714]
When attempting to build my app I got the following error:

```
Launching lib\main.dart on SM A705FN in debug mode...
Running Gradle task 'assembleDebug'...
E:\flutter\flutter_course\xylophone\android\app\src\debug\AndroidManifest.xml Error:
	uses-sdk:minSdkVersion 16 cannot be smaller than version 23 declared in library [:audioplayers] E:\flutter\flutter_course\xylophone\build\audioplayers\intermediates\library_manifest\debug\AndroidManifest.xml as the library might be using APIs not available in 16
	Suggestion: use a compatible library with a minSdk of at most 16,
		or increase this project's minSdk version to at least 23,
		or use tools:overrideLibrary="xyz.luan.audioplayers" to force usage (may lead to runtime failures)

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:processDebugMainManifest'.
> Manifest merger failed : uses-sdk:minSdkVersion 16 cannot be smaller than version 23 declared in library [:audioplayers] E:\flutter\flutter_course\xylophone\build\audioplayers\intermediates\library_manifest\debug\AndroidManifest.xml as the library might be using APIs not available in 16
  	Suggestion: use a compatible library with a minSdk of at most 16,
  		or increase this project's minSdk version to at least 23,
  		or use tools:overrideLibrary="xyz.luan.audioplayers" to force usage (may lead to runtime failures)

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 16s
Exception: Gradle task assembleDebug failed with exit code 1
```

So, I went into `android\app\build.gradle` and changed the value of `minSdkVersion` from `16` to `23`. I then get this error instead:

```
Launching lib\main.dart on SM A705FN in debug mode...
Running Gradle task 'assembleDebug'...
e: C:\Flutter\.pub-cache\hosted\pub.dartlang.org\audioplayers-0.17.1\android\src\main\kotlin\xyz\luan\audioplayers\AudioplayersPlugin.kt: (181, 52): Expecting a parameter declaration
e: C:\Flutter\.pub-cache\hosted\pub.dartlang.org\audioplayers-0.17.1\android\src\main\kotlin\xyz\luan\audioplayers\AudioplayersPlugin.kt: (231, 38): Expecting an argument
e: C:\Flutter\.pub-cache\hosted\pub.dartlang.org\audioplayers-0.17.1\android\src\main\kotlin\xyz\luan\audioplayers\ByteDataSource.kt: (8, 37): Expecting a parameter declaration
e: C:\Flutter\.pub-cache\hosted\pub.dartlang.org\audioplayers-0.17.1\android\src\main\kotlin\xyz\luan\audioplayers\WrappedMediaPlayer.kt: (10, 39): Expecting a parameter declaration
e: C:\Flutter\.pub-cache\hosted\pub.dartlang.org\audioplayers-0.17.1\android\src\main\kotlin\xyz\luan\audioplayers\WrappedSoundPool.kt: (168, 32): Expecting a parameter declaration
e: C:\Flutter\.pub-cache\hosted\pub.dartlang.org\audioplayers-0.17.1\android\src\main\kotlin\xyz\luan\audioplayers\WrappedSoundPool.kt: (205, 26): Expecting an argument
e: C:\Flutter\.pub-cache\hosted\pub.dartlang.org\audioplayers-0.17.1\android\src\main\kotlin\xyz\luan\audioplayers\WrappedSoundPool.kt: (46, 77): Type inference failed. Expected type mismatch: inferred type is List<???> but MutableList<WrappedSoundPool> was expected

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':audioplayers:compileDebugKotlin'.
> Compilation error. See log for more details

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 18s
Exception: Gradle task assembleDebug failed with exit code 1
```

**Comment by darre1** [comment]
FYI - If (after doing the above) I downgrade the package from 0.17.1 to 0.17.0 my app then compiles and produces the following warnings: 

```
Launching lib\main.dart on SM A705FN in debug mode...
Running Gradle task 'assembleDebug'...
C:\Flutter\.pub-cache\hosted\pub.dartlang.org\audioplayers-0.17.0\android\src\main\java\xyz\luan\audioplayers\AudioplayersPlugin.java:18: warning: [deprecation] Registrar in PluginRegistry has been deprecated
import io.flutter.plugin.common.PluginRegistry.Registrar;
                                              ^
C:\Flutter\.pub-cache\hosted\pub.dartlang.org\audioplayers-0.17.0\android\src\main\java\xyz\luan\audioplayers\AudioplayersPlugin.java:26: warning: [deprecation] Handler() in Handler has been deprecated
    private final Handler handler = new Handler();
                                    ^
C:\Flutter\.pub-cache\hosted\pub.dartlang.org\audioplayers-0.17.0\android\src\main\java\xyz\luan\audioplayers\AudioplayersPlugin.java:31: warning: [deprecation] Registrar in PluginRegistry has been deprecated
    public static void registerWith(final Registrar registrar) {
                                          ^
3 warnings
√ Built build\app\outputs\flutter-apk\app-debug.apk.
Installing build\app\outputs\flutter-apk\app.apk...
```

**Comment by alainseys** [comment]
i got the same issue, reverting to an older version of audioplayers does not solve it.

  audioplayers: ^0.17.1


`alainseys@MacBook-Pro-van-Alain:~/Documents/Development Projects/Flutter/DoobeeApp (remotefunctions*=) % flutter run --debug --enable-software-rendering
Using software rendering with device Android SDK built for x86. You may get better performance with hardware mode by configuring hardware rendering for your device.
Launching lib/main.dart on Android SDK built for x86 in debug mode...
e: /Users/alainseys/.pub-cache/hosted/pub.dartlang.org/audioplayers-0.17.1/android/src/main/kotlin/xyz/luan/audioplayers/AudioplayersPlugin.kt: (181, 52): Expecting a parameter declaration
e: /Users/alainseys/.pub-cache/hosted/pub.dartlang.org/audioplayers-0.17.1/android/src/main/kotlin/xyz/luan/audioplayers/AudioplayersPlugin.kt: (231, 38): Expecting an argument
e: /Users/alainseys/.pub-cache/hosted/pub.dartlang.org/audioplayers-0.17.1/android/src/main/kotlin/xyz/luan/audioplayers/ByteDataSource.kt: (8, 37): Expecting a parameter declaration
e: /Users/alainseys/.pub-cache/hosted/pub.dartlang.org/audioplayers-0.17.1/android/src/main/kotlin/xyz/luan/audioplayers/WrappedMediaPlayer.kt: (10, 39): Expecting a parameter declaration
e: /Users/alainseys/.pub-cache/hosted/pub.dartlang.org/audioplayers-0.17.1/android/src/main/kotlin/xyz/luan/audioplayers/WrappedSoundPool.kt: (168, 32): Expecting a parameter declaration
e: /Users/alainseys/.pub-cache/hosted/pub.dartlang.org/audioplayers-0.17.1/android/src/main/kotlin/xyz/luan/audioplayers/WrappedSoundPool.kt: (205, 26): Expecting an argument
e: /Users/alainseys/.pub-cache/hosted/pub.dartlang.org/audioplayers-0.17.1/android/src/main/kotlin/xyz/luan/audioplayers/WrappedSoundPool.kt: (46, 77): Type inference failed. Expected type mismatch: inferred type is List<???> but MutableList<WrappedSoundPool> was expected
`

**Comment by another-guy** [comment]
@alainseys  I made a mistake when downgrading to 0.17.0. Try again and point to the specific version as `audioplayers: 0.17.0`. Make sure to NOT have the `^` symbol before the version. Maybe it will work for you too.

**Comment by huantd0111** [comment]
@darre1  @alainseys  you can try updagrade ext.kotlin_version to  '1.4.21' on build.gradle.
It work for me!

**Comment by alainseys** [comment]
@huantd0111  @another-guy  Thanks for the info this solved the issue for me

**Comment by jinfagang** [comment]
@luanpotter  Would original author make sure lib can be correct compile before upgrade a new version? Otherwise it confuse users when using it...

**Comment by erickzanardo** [comment]
@darre1 as your log shows, your app is using a min sdk version lower than what supported by Audioplayers. On the last version the min sdk version was increase to 23, you will need to increase the min sdk version of your application.

About the other issues reported here, they doesn't seem to be related to the one reported by the author of this issue, so feel free to open specific issues for those cases.

**Comment by darre1** [comment]
> @darre1 as your log shows, your app is using a min sdk version lower than what supported by Audioplayers. On the last version the min sdk version was increase to 23, you will need to increase the min sdk version of your application.

I tried that, which just produced a different error - I stated that in my original post too...

**Comment by erickzanardo** [comment]
@darre1 alright! my bad I didn't noticed the second part of the issue, I will reopen this, but could you edit and give focus on the second error please?

**Comment by darre1** [comment]
> @darre1 alright! my bad I didn't noticed the second part of the issue, I will reopen this, but could you edit and give focus on the second error please?

ok, here you go...


v1.17.1 of this package requires a minimum SDK version of 23 so I went into `android\app\build.gradle` and changed the value of `minSdkVersion` from `16` to `23`. I then get this error:

```
Launching lib\main.dart on SM A705FN in debug mode...
Running Gradle task 'assembleDebug'...
e: C:\Flutter\.pub-cache\hosted\pub.dartlang.org\audioplayers-0.17.1\android\src\main\kotlin\xyz\luan\audioplayers\AudioplayersPlugin.kt: (181, 52): Expecting a parameter declaration
e: C:\Flutter\.pub-cache\hosted\pub.dartlang.org\audioplayers-0.17.1\android\src\main\kotlin\xyz\luan\audioplayers\AudioplayersPlugin.kt: (231, 38): Expecting an argument
e: C:\Flutter\.pub-cache\hosted\pub.dartlang.org\audioplayers-0.17.1\android\src\main\kotlin\xyz\luan\audioplayers\ByteDataSource.kt: (8, 37): Expecting a parameter declaration
e: C:\Flutter\.pub-cache\hosted\pub.dartlang.org\audioplayers-0.17.1\android\src\main\kotlin\xyz\luan\audioplayers\WrappedMediaPlayer.kt: (10, 39): Expecting a parameter declaration
e: C:\Flutter\.pub-cache\hosted\pub.dartlang.org\audioplayers-0.17.1\android\src\main\kotlin\xyz\luan\audioplayers\WrappedSoundPool.kt: (168, 32): Expecting a parameter declaration
e: C:\Flutter\.pub-cache\hosted\pub.dartlang.org\audioplayers-0.17.1\android\src\main\kotlin\xyz\luan\audioplayers\WrappedSoundPool.kt: (205, 26): Expecting an argument
e: C:\Flutter\.pub-cache\hosted\pub.dartlang.org\audioplayers-0.17.1\android\src\main\kotlin\xyz\luan\audioplayers\WrappedSoundPool.kt: (46, 77): Type inference failed. Expected type mismatch: inferred type is List<???> but MutableList<WrappedSoundPool> was expected

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':audioplayers:compileDebugKotlin'.
> Compilation error. See log for more details

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 18s
Exception: Gradle task assembleDebug failed with exit code 1
```

**Comment by sai-pavan-tej** [comment]
I do follow same steps for that issue...
i'm also got the same error....

> FAILURE: Build failed with an exception.
> 
> * What went wrong:
> Execution failed for task ':audioplayers:compileDebugKotlin'.
> > Compilation error. See log for more details
> 
> * Try:
> Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to 
> get full insights.
> 
> * Get more help at https://help.gradle.org
> 
> BUILD FAILED in 10s
> Running Gradle task 'assembleDebug'...                                  
> Running Gradle task 'assembleDebug'... Done                        11.3s
> Exception: Gradle task assembleDebug failed with exit code 1

**Comment by pradeepsthapa** [comment]
go to AndroidManifest.xml and add the line
`<uses-sdk android:minSdkVersion="16" android:targetSdkVersion="30" tools:overrideLibrary="xyz.luan.audioplayers"/>`
alse add `xmlns:tools="http://schemas.android.com/tools">` after package name.
This worked for me.

**Comment by sai-pavan-tej** [comment]
it's automatically come back to older code(in AndroidManifest.xml ) when i build.

**Comment by pradeepsthapa** [comment]
> it's automatically come back to older code(in AndroidManifest.xml ) when i build.

did you changed the AndroidManifest.xml file of android/src/main/AndroidManifest.xml or android/src/debug/AndroidManifest.xml . You should change the main.

**Comment by sai-pavan-tej** [comment]
Ohh my bad I did change in audioplayer's  AndroidManifest.xml file

**Comment by sai-pavan-tej** [comment]
this is how paste the above two lines.. is it correct?
Actually it throws so many errors..

> <manifest xmlns:android="http://schemas.android.com/apk/res/android"
> 
>     package="com.example.spanishnumberaudio" 
>     xmlns:tools="http://schemas.android.com/tools">
>     
>     <application
>         android:name="io.flutter.app.FlutterApplication"
>         android:label="spanishnumberaudio"
>         android:icon="@mipmap/ic_launcher">
>         <activity
>             android:name=".MainActivity"
>             android:launchMode="singleTop"
>             android:theme="@style/LaunchTheme"
>             android:configChanges="orientation|keyboardHidden|keyboard|screenSize|smallestScreenSize|locale|layoutDirection|fontScale|screenLayout|density|uiMode"
>             android:hardwareAccelerated="true"
>             android:windowSoftInputMode="adjustResize">
>             
> 
>             <meta-data
>               android:name="io.flutter.embedding.android.NormalTheme"
>               android:resource="@style/NormalTheme"
>               />
>             
>             <meta-data
>               android:name="io.flutter.embedding.android.SplashScreenDrawable"
>               android:resource="@drawable/launch_background"
>               />
>             <intent-filter>
>                 <action android:name="android.intent.action.MAIN"/>
>                 <category android:name="android.intent.category.LAUNCHER"/>
>             </intent-filter>
>         </activity>
>         
>         <meta-data
>             android:name="flutterEmbedding"
>             android:value="2" />
> 
>     </application>
>     <uses-sdk android:minSdkVersion="16" android:targetSdkVersion="30" tools:overrideLibrary="xyz.luan.audioplayers"/>
> </manifest>
>

**Comment by pradeepsthapa** [comment]
> this is how paste the above two lines.. is it correct?
> Actually it throws so many errors..
> 
> > <manifest xmlns:android="http://schemas.android.com/apk/res/android"
> > ```
> > package="com.example.spanishnumberaudio" 
> > xmlns:tools="http://schemas.android.com/tools">
> > 
> > <application
> >     android:name="io.flutter.app.FlutterApplication"
> >     android:label="spanishnumberaudio"
> >     android:icon="@mipmap/ic_launcher">
> >     <activity
> >         android:name=".MainActivity"
> >         android:launchMode="singleTop"
> >         android:theme="@style/LaunchTheme"
> >         android:configChanges="orientation|keyboardHidden|keyboard|screenSize|smallestScreenSize|locale|layoutDirection|fontScale|screenLayout|density|uiMode"
> >         android:hardwareAccelerated="true"
> >         android:windowSoftInputMode="adjustResize">
> >         
> > 
> >         <meta-data
> >           android:name="io.flutter.embedding.android.NormalTheme"
> >           android:resource="@style/NormalTheme"
> >           />
> >         
> >         <meta-data
> >           android:name="io.flutter.embedding.android.SplashScreenDrawable"
> >           android:resource="@drawable/launch_background"
> >           />
> >         <intent-filter>
> >             <action android:name="android.intent.action.MAIN"/>
> >             <category android:name="android.intent.category.LAUNCHER"/>
> >         </intent-filter>
> >     </activity>
> >     
> >     <meta-data
> >         android:name="flutterEmbedding"
> >         android:value="2" />
> > 
> > </application>
> > <uses-sdk android:minSdkVersion="16" android:targetSdkVersion="30" tools:overrideLibrary="xyz.luan.audioplayers"/>
> > ```

add before application tag like this.
`<manifest xmlns:android="http://schemas.android.com/apk/res/android"
    package="com.example.spanishnumberaudio"
    xmlns:tools="http://schemas.android.com/tools">
    <uses-permission android:name="android.permission.INTERNET" />
    <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
    <uses-sdk android:minSdkVersion="16" android:targetSdkVersion="30" tools:overrideLibrary="xyz.luan.audioplayers"/>
    <application
        android:name="io.flutter.app.FlutterApplication"`....

**Comment by rwrz** [comment]
I just sent the fixes needed to compile it again. Was just some kotlin sintax problems.

**Comment by SkilyFrance** [comment]
Hey i met the same problem, **upgrade your minSdkVersion from 16 to 23 in build.gradle** and follow these steps :

1. Build without audioplayers package
- Remove audioplayers package from your pubspec.yaml
- Flutter clean + invalidate cache + wipe data (emulator)
- Flutter build

2. Rebuild with audioplayers package
- re-implement audioplayers package on pubspec.yaml
- run Pub get packages
- Flutter build

This kind of log "Execution failed for task ':audioplayers:compileDebugKotlin'.", sometimes is caused by .pubcache folder. So to clean it, follow the above process. Then, you can restart on clean bases.

If nothing happens check if your Android Gradle plugin matchs with your version of gradle thanks o this link :
https://developer.android.com/studio/releases/gradle-plugin

Hope it's helpful champ,
Guillaume

**Comment by benzhir** [comment]
@huantd0111 Thanks for the info this solved the issue for me

**Comment by i3games** [comment]
>     1. Build without audioplayers package
[...,]

did not work for me as did none of the othe rsuggestions in this thread.  
I also tried `pub cache repair` and updated the latest Kotlin plugin 1.4.21

I am still getting 

```
e: /Users/xxx/root/topics/coding/flutter/sdk/.pub-cache/hosted/pub.dartlang.org/audioplayers-0.17.2/android/src/main/kotlin/xyz/luan/audioplayers/AudioplayersPlugin.kt: (181, 52): Expecting a parameter declaration
e: /Users/xxx/root/topics/coding/flutter/sdk/.pub-cache/hosted/pub.dartlang.org/audioplayers-0.17.2/android/src/main/kotlin/xyz/luan/audioplayers/AudioplayersPlugin.kt: (231, 38): Expecting an argument
e: /Users/xxx/root/topics/coding/flutter/sdk/.pub-cache/hosted/pub.dartlang.org/audioplayers-0.17.2/android/src/main/kotlin/xyz/luan/audioplayers/ByteDataSource.kt: (8, 37): Expecting a parameter declaration
e: /Users/xxx/root/topics/coding/flutter/sdk/.pub-cache/hosted/pub.dartlang.org/audioplayers-0.17.2/android/src/main/kotlin/xyz/luan/audioplayers/WrappedMediaPlayer.kt: (10, 39): Expecting a parameter declaration
e: /Users/xxx/root/topics/coding/flutter/sdk/.pub-cache/hosted/pub.dartlang.org/audioplayers-0.17.2/android/src/main/kotlin/xyz/luan/audioplayers/WrappedSoundPool.kt: (168, 32): Expecting a parameter declaration
e: /Users/xxx/root/topics/coding/flutter/sdk/.pub-cache/hosted/pub.dartlang.org/audioplayers-0.17.2/android/src/main/kotlin/xyz/luan/audioplayers/WrappedSoundPool.kt: (205, 26): Expecting an argument
e: /Users/xxx/root/topics/coding/flutter/sdk/.pub-cache/hosted/pub.dartlang.org/audioplayers-0.17.2/android/src/main/kotlin/xyz/luan/audioplayers/WrappedSoundPool.kt: (46, 77): Type inference failed. Expected type mismatch: inferred type is List<???> but MutableList<WrappedSoundPool> was expected

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':audioplayers:compileDebugKotlin'.
> Compilation error. See log for more details

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 7s
Exception: Gradle task assembleDebug failed with exit code 1
```

audioplayers 0.17.2, physical device. The project compiles on iOS.

I would also second asking for more testing before releasing new versions.

**Comment by dokind** [comment]
tried all things but the result is:
E/MediaPlayerNative(17004): error (1, -2147483648)
E/MediaPlayer(17004): Error (1,-2147483648)
E/MediaPlayerNative(17004): pause called in state 0, mPlayer(0xd1f85120)
E/MediaPlayerNative(17004): error (-38, 0)
E/MediaPlayerNative(17004): Attempt to perform seekTo in wrong state: mPlayer=0xd1f85120, mCurrentState=0
E/MediaPlayerNative(17004): error (-38, 0)
2
E/MediaPlayer(17004): Error (-38,0)

**Comment by NatnaelAlemayehu** [comment]
I had the same exact issue. Just change the audioplayer version to a lower one. For me I used 0.8.0 ... worked like a charm. Don't forget to pub get after the changing the version number though.

**Comment by luanpotter** [comment]
To summarize this issue:

The error is your project is using an outdated version of kotlin that didn't support some features like trailing commas.
The real fix is to update your app (explained above).
However to avoid confusion I merged [this PR](https://github.com/luanpotter/audioplayers/pull/732) that also downgrades our code so that it works on old versions of kotlin. That has been merged and released on 0.17.3 and up (currently we are at 0.17.4).
Therefore, this is fixed. There is no reason to downgrade AP anymore on your apps.
